File: OJ practices/LeetCode题目集/88. Merge Sorted Array(easy).cpp
/**
 * @brief 合并有序数组，逆向双指针法
 * 时间复杂度 O(m+n) 空间复杂度 O(1)
 * 
 */

/*
https://leetcode.com/problems/merge-sorted-array/
https://leetcode.cn/problems/merge-sorted-array/solutions/666608/he-bing-liang-ge-you-xu-shu-zu-by-leetco-rrb0/?envType=study-plan-v2&envId=top-interview-150
 */

#include <vector>

using namespace std;

class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {
        int i = m - 1;
        int j = n - 1;
        int k = m + n - 1;
        while (0 <= i && 0 <= j)
        {
            if (nums1[i] < nums2[j])
            {
                nums1[k] = nums2[j];
                --j;
            }
            else
            {
                nums1[k] = nums1[i];
                --i;
            }
            --k;
        }
        while (0 <= i)
        {
            nums1[k--] = nums1[i--];
        }
        while (0 <= j)
        {
            nums1[k--] = nums2[j--];
        }
        
    }
};

File: OJ practices/LeetCode题目集/88. Merge Sorted Array(easy).py
'''
@Author: SourDumplings
@Date: 2020-02-16 19:37:18
@Link: https://github.com/SourDumplings/
@Email: changzheng300@foxmail.com
@Description: https://leetcode-cn.com/problems/merge-sorted-array/

双指针法，从后往前填
'''

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        p = m + n - 1
        p1 = m - 1
        p2 = n - 1
        while 0 <= p1 and 0 <= p2:
            num1 = nums1[p1]
            num2 = nums2[p2]
            if num1 < num2:
                nums1[p] = nums2[p2]
                p2 -= 1
            else:
                nums1[p] = nums1[p1]
                p1 -= 1
            p -= 1

        while 0 <= p and p1 == -1:
            nums1[p] = nums2[p2]
            p -= 1
            p2 -= 1
        while 0 <= p and p2 == -1:
            nums1[p] = nums1[p1]
            p -= 1
            p1 -= 1
            
